test/unit/org/apache/cassandra/db/PartitionRangeReadTest.java

```diff
@@ -188,5 +201,94 @@ public void testRangeSliceInclusionExclusion() throws Throwable
+        assertEquals(20, cf); // because 100-80 = 20 < maxConccurrentRangeRequest
+
+        // returned half rows, fetch rangesQueried again but hit the max instead
+        cf = StorageProxy.RangeCommandIterator.computeConcurrencyFactor(100, 60, maxConccurrentRangeRequest, 480, 240);
+        assertEquals(maxConccurrentRangeRequest, cf); // because 60 > maxConccurrentRangeRequest
+
+        // returned half rows, fetch rangesQueried again
+        cf = StorageProxy.RangeCommandIterator.computeConcurrencyFactor(100, 30, maxConccurrentRangeRequest, 480, 240);
+        assertEquals(30, cf); // because 30 < maxConccurrentRangeRequest
+
+        // returned most of rows, 1 more range to fetch
+        cf = StorageProxy.RangeCommandIterator.computeConcurrencyFactor(100, 1, maxConccurrentRangeRequest, 480, 479);
+        assertEquals(1, cf); // because 1 < maxConccurrentRangeRequest
+    }
+
+    @Test
+    public void testRangeCountWithRangeMerge()
+    {
+        List<Token> tokens = updateTokens(Arrays.asList(100, 200, 300, 400));
+
+        Keyspace keyspace = Keyspace.open(KEYSPACE1);
+        List<ReplicaPlan.ForRangeRead> ranges = new ArrayList<>();
+        for (int i = 0; i + 1 < tokens.size(); i++)
+        {
+            Range<PartitionPosition> range = Range.makeRowRange(tokens.get(i), tokens.get(i + 1));
+            ranges.add(ReplicaPlans.forRangeRead(keyspace, ConsistencyLevel.ONE, range, 1));
+        }
+
+        StorageProxy.RangeMerger merge = new StorageProxy.RangeMerger(ranges.iterator(), keyspace, ConsistencyLevel.ONE);
+        ReplicaPlan.ForRangeRead mergedRange = Iterators.getOnlyElement(merge);
+        // all ranges are merged as test has only one node.
+        assertEquals(tokens.size() - 1, mergedRange.vnodeCount());
+    }
+
+    @Test
+    public void testRangeQueried()
+    {
+        List<Token> tokens = updateTokens(Arrays.asList(100, 200, 300, 400));
+
+        Keyspace keyspace = Keyspace.open(KEYSPACE1);
+        ColumnFamilyStore cfs = keyspace.getColumnFamilyStore(CF_STANDARD1);
+        cfs.clearUnsafe();
+
+        int rows = 100;
+        for (int i = 0; i < rows; ++i)
+        {
+            RowUpdateBuilder builder = new RowUpdateBuilder(cfs.metadata(), 10, String.valueOf(i));
+            builder.clustering("c");
+            builder.add("val", String.valueOf(i));
+            builder.build().applyUnsafe();
+        }
+
+        cfs.forceBlockingFlush();
+
+        PartitionRangeReadCommand command = (PartitionRangeReadCommand) Util.cmd(cfs).build();
+        // avoid merging ranges, so that it queries in multiple batches and check if liveReturned is updated correctly.
+        StorageProxy.RangeIterator rangeIterator = new StorageProxy.RangeIterator(command, keyspace, ConsistencyLevel.ONE);
+        StorageProxy.RangeCommandIterator data = new StorageProxy.RangeCommandIterator(rangeIterator, command, 1, 1000, keyspace, ConsistencyLevel.ONE, System.nanoTime());
```

Comment:
@jasonstack I traced through this test and it actually looks like we *are* merging all 5 ranges into one `ForRangeRead`, and the reason we still get 5 from `data.rangesQueried()` is just that we allow overflow from the _last_ `ForRangeRead` in `sendNextRequests()`. It feels like perhaps a couple more tests around `sendNextRequests()` would be helpful? If we parameterize `RangeCommandIterator` to _optionally_ merge ranges, this would be easier. (It could even make the signature of the `RangeCommandIterator` constructor less busy, given `keyspace` and `consistency` are actually only used by the `RangeMerger` created inside it.)

Reply:
Renamed `data.rangesQueried()` to `data.vnodeRangesQueried()`.. I will add some more tests